Improve performance and output of `pyenv virtualenvs`

samdoran opened the pull request:
The biggest change is reusing most of the code used for `pyenv versions` in `pyenv virtualenvs`. This dramatically improves performance while making the output consistent.

Related to #490. Maybe fixes the problem?

With 128 virtual environments, it currently takes about 13 seconds to run `pyenv virtualenvs --bare`.

```shell
> time pyenv virtualenvs --bare > /dev/null 2>&1

________________________________________________________
Executed in   13.14 secs    fish           external
   usr time    6.32 secs    0.33 millis    6.32 secs
   sys time    6.13 secs    2.19 millis    6.13 secs
```

<details>
  <summary>Output</summary>
  
```
1brc
3.10.13/envs/instruct-lab
3.10.13/envs/parquet-compactor-3.10.13
3.10.14/envs/ipython-3.10.14
3.10.15/envs/ipython-3.10.15
3.10.7/envs/cherry_picker-3.10.7
3.10.9/envs/pyls-3.10.9
3.11.0/envs/argcomplete-3.11.0
3.11.0/envs/asciinema-3.11.0
3.11.0/envs/borg-3.11.0
3.11.0/envs/cherry_picker-3.11.0
3.11.0/envs/cookiecutter-3.11.0
3.11.0/envs/diceware-3.11.0
3.11.0/envs/flake8-3.11.0
3.11.0/envs/ipdb-3.11.0
3.11.0/envs/isort-3.11.0
3.11.0/envs/jello-3.11.0
3.11.0/envs/mackup-3.11.0
3.11.0/envs/maestral-3.11.0
3.11.0/envs/mitmproxy-3.11.0
3.11.0/envs/mypy-3.11.0
3.11.0/envs/pipfile2req-3.11.0
3.11.0/envs/pur-3.11.0
3.11.0/envs/pylint-3.11.0
3.11.0/envs/pyspy-3.11.0
3.11.0/envs/pytest-3.11.0
3.11.0/envs/qrimg-3.11.0
3.11.0/envs/shiv-3.11.0
3.11.0/envs/speedtest-cli-3.11.0
3.11.0/envs/sphinx-3.11.0
3.11.0/envs/tox-3.11.0
3.11.0/envs/tweet-delete-3.11.0
3.11.0/envs/yamllint-3.11.0
3.11.0/envs/youtube-dl-3.11.0
3.11.1/envs/borg-3.11.1
3.11.1/envs/qontract-3.11.1
3.11.1/envs/requests-3.11.1
3.11.1/envs/titlecase-3.11.1
3.11.1/envs/uptime-kuma-3.11.1
3.11.10/envs/ipython-3.11.10
3.11.10/envs/koku-dev-3.11.10
3.11.10/envs/koku-test
3.11.2/envs/bonfire-3.11.2
3.11.2/envs/testing
3.11.3/envs/ansible-dev-3.11.3
3.11.3/envs/qr-coding-3.11.3
3.11.4/envs/flash
3.11.4/envs/json-linter-3.11.4
3.11.4/envs/youtube-dl-dev
3.11.5/envs/parquet-compactor-3.11.5
3.11.6/envs/ansible-core-3.11.6
3.11.6/envs/pre-commit
3.11.6/envs/pylibssh-dev-3.11.6
3.11.6/envs/tasmota-3.11.6
3.11.7/envs/pipenv-3.11.7
3.11.7/envs/tasmota-3.11.7
3.11.8/envs/cffi-dev-3.11.8
3.11.8/envs/koku-dev-3.11.8
3.11.8/envs/koku-dev-dj4-3.11.8
3.11.9/envs/ipython-3.11.9
3.11.9/envs/koku-dev-3.11.9
3.11.9/envs/nise-populator-3.11.9
3.12.0/envs/ansible-core-3.12.0
3.12.0/envs/cla-dev-3.12.0
3.12.0/envs/ipython-3.12.0
3.12.0/envs/jupyter
3.12.0/envs/koku-metrics-operator-3.12.0
3.12.0/envs/pytorch
3.12.0/envs/requests-3.12.0
3.12.0/envs/rh-pre-commit-3.12.0
3.12.1/envs/ipython-3.12.1
3.12.2/envs/1brc
3.12.2/envs/ansible-core-3.12.2
3.12.2/envs/black
3.12.2/envs/cffi-dev-3.12.2
3.12.2/envs/codecov-cli
3.12.2/envs/dr-content-3.12.2
3.12.2/envs/harlequin
3.12.2/envs/ipython-3.12.2
3.12.2/envs/pip-audit
3.12.2/envs/pygments-3.12.2
3.12.2/envs/pyright
3.12.2/envs/qrcode
3.12.2/envs/ruff
3.12.3/envs/ipython-3.12.3
3.12.5/envs/ipython-3.12.5
3.12.5/envs/shellai
3.12.6/envs/build-definitions
3.12.6/envs/ipython-3.12.6
3.12.6/envs/kflux-rel-data
3.12.7/envs/aoc-2024
3.12.7/envs/ipython-3.12.7
3.12.8/envs/dr-backend-3.12.8
3.12.8/envs/ipython-3.12.8
3.12.8/envs/lifecycle-defs
3.12.8/envs/rlsapi-dev
3.12.8/envs/rlsrag
3.13.0/envs/cffi-3.13.0
3.13.0/envs/cicd-tools
3.13.0/envs/ipython-3.13.0
3.13.0/envs/thefuck
3.13.1/envs/docling
3.13.1/envs/httpie
3.13.1/envs/ipython-3.13.1
3.13.1/envs/llm
3.13.1/envs/locust
3.13.1/envs/poetry
3.13.1/envs/prefix
3.13.1/envs/tox
3.6.15/envs/ipython-3.6.15
3.7.15/envs/ipython-3.7.15
3.7.16/envs/ipython-3.7.16
3.8.14/envs/ansible-dev-3.8.14
3.8.15/envs/ipython-3.8.15
3.8.15/envs/nise-populator-3.8.15
3.8.16/envs/koku-nise-dev-3.8.16
3.8.17/envs/ipython-3.8.17
3.8.18/envs/parquet-compactor-3.8.18
3.8.20/envs/ipython-3.8.20
3.9.14/envs/iqe-3.9.14
3.9.15/envs/ipython-3.9.15
3.9.16/envs/koku-dev-logging-3.9.16
3.9.17/envs/event-sign-3.9.17
3.9.17/envs/ipython-3.9.17
3.9.18/envs/koku-dev-3.9.18
3.9.18/envs/koku-rediscache-debug
3.9.18/envs/parquet-compactor-3.9.18
3.9.19/envs/ipython-3.9.19
3.9.19/envs/koku-dev-3.9.19
ansible-core-3.11.6
ansible-core-3.12.0
ansible-core-3.12.2
ansible-dev-3.11.3
ansible-dev-3.8.14
aoc-2024
argcomplete-3.11.0
asciinema-3.11.0
black
bonfire-3.11.2
borg-3.11.0
borg-3.11.1
build-definitions
cffi-3.13.0
cffi-dev-3.11.8
cffi-dev-3.12.2
cherry_picker-3.10.7
cherry_picker-3.11.0
cicd-tools
cla-dev-3.12.0
codecov-cli
cookiecutter-3.11.0
diceware-3.11.0
docling
dr-backend-3.12.8
dr-content-3.12.2
event-sign-3.9.17
flake8-3.11.0
flash
harlequin
httpie
instruct-lab
ipdb-3.11.0
ipython-3.10.14
ipython-3.10.15
ipython-3.11.10
ipython-3.11.9
ipython-3.12.0
ipython-3.12.1
ipython-3.12.2
ipython-3.12.3
ipython-3.12.5
ipython-3.12.6
ipython-3.12.7
ipython-3.12.8
ipython-3.13.0
ipython-3.13.1
ipython-3.6.15
ipython-3.7.15
ipython-3.7.16
ipython-3.8.15
ipython-3.8.17
ipython-3.8.20
ipython-3.9.15
ipython-3.9.17
ipython-3.9.19
iqe-3.9.14
isort-3.11.0
jello-3.11.0
json-linter-3.11.4
jupyter
kflux-rel-data
koku-dev-3.11.10
koku-dev-3.11.8
koku-dev-3.11.9
koku-dev-3.9.18
koku-dev-3.9.19
koku-dev-dj4-3.11.8
koku-dev-logging-3.9.16
koku-metrics-operator-3.12.0
koku-nise-dev-3.8.16
koku-rediscache-debug
koku-test
lifecycle-defs
llm
locust
mackup-3.11.0
maestral-3.11.0
mitmproxy-3.11.0
mypy-3.11.0
nise-populator-3.11.9
nise-populator-3.8.15
parquet-compactor-3.10.13
parquet-compactor-3.11.5
parquet-compactor-3.8.18
parquet-compactor-3.9.18
pip-audit
pipenv-3.11.7
pipfile2req-3.11.0
poetry
pre-commit
prefix
pur-3.11.0
pygments-3.12.2
pylibssh-dev-3.11.6
pylint-3.11.0
pyls-3.10.9
pyright
pyspy-3.11.0
pytest-3.11.0
pytorch
qontract-3.11.1
qr-coding-3.11.3
qrcode
qrimg-3.11.0
requests-3.11.1
requests-3.12.0
rh-pre-commit-3.12.0
rlsapi-dev
rlsrag
ruff
shellai
shiv-3.11.0
speedtest-cli-3.11.0
sphinx-3.11.0
tasmota-3.11.6
tasmota-3.11.7
testing
thefuck
titlecase-3.11.1
tox
tox-3.11.0
tweet-delete-3.11.0
uptime-kuma-3.11.1
yamllint-3.11.0
youtube-dl-3.11.0
youtube-dl-dev

```
</details>

With the changes in this branch it is 2.9s:
```shell
> time pyenv virtualenvs --bare > /dev/null 2>&1

________________________________________________________
Executed in    2.90 secs    fish           external
   usr time    1.28 secs    0.31 millis    1.28 secs
   sys time    1.38 secs    1.29 millis    1.38 secs
```

<details>
  <summary>Output</summary>
  
```
3.10.13/envs/instruct-lab
3.10.13/envs/parquet-compactor-3.10.13
3.10.14/envs/ipython-3.10.14
3.10.15/envs/ipython-3.10.15
3.10.7/envs/cherry_picker-3.10.7
3.10.9/envs/pyls-3.10.9
3.11.0/envs/argcomplete-3.11.0
3.11.0/envs/asciinema-3.11.0
3.11.0/envs/borg-3.11.0
3.11.0/envs/cherry_picker-3.11.0
3.11.0/envs/cookiecutter-3.11.0
3.11.0/envs/diceware-3.11.0
3.11.0/envs/flake8-3.11.0
3.11.0/envs/ipdb-3.11.0
3.11.0/envs/isort-3.11.0
3.11.0/envs/jello-3.11.0
3.11.0/envs/mackup-3.11.0
3.11.0/envs/maestral-3.11.0
3.11.0/envs/mitmproxy-3.11.0
3.11.0/envs/mypy-3.11.0
3.11.0/envs/pipfile2req-3.11.0
3.11.0/envs/pur-3.11.0
3.11.0/envs/pylint-3.11.0
3.11.0/envs/pyspy-3.11.0
3.11.0/envs/pytest-3.11.0
3.11.0/envs/qrimg-3.11.0
3.11.0/envs/shiv-3.11.0
3.11.0/envs/speedtest-cli-3.11.0
3.11.0/envs/sphinx-3.11.0
3.11.0/envs/tox-3.11.0
3.11.0/envs/tweet-delete-3.11.0
3.11.0/envs/yamllint-3.11.0
3.11.0/envs/youtube-dl-3.11.0
3.11.1/envs/borg-3.11.1
3.11.1/envs/qontract-3.11.1
3.11.1/envs/requests-3.11.1
3.11.1/envs/titlecase-3.11.1
3.11.1/envs/uptime-kuma-3.11.1
3.11.10/envs/ipython-3.11.10
3.11.10/envs/koku-dev-3.11.10
3.11.10/envs/koku-test
3.11.2/envs/bonfire-3.11.2
3.11.2/envs/testing
3.11.3/envs/ansible-dev-3.11.3
3.11.3/envs/qr-coding-3.11.3
3.11.4/envs/flash
3.11.4/envs/json-linter-3.11.4
3.11.4/envs/youtube-dl-dev
3.11.5/envs/parquet-compactor-3.11.5
3.11.6/envs/ansible-core-3.11.6
3.11.6/envs/pre-commit
3.11.6/envs/pylibssh-dev-3.11.6
3.11.6/envs/tasmota-3.11.6
3.11.7/envs/pipenv-3.11.7
3.11.7/envs/tasmota-3.11.7
3.11.8/envs/cffi-dev-3.11.8
3.11.8/envs/koku-dev-3.11.8
3.11.8/envs/koku-dev-dj4-3.11.8
3.11.9/envs/ipython-3.11.9
3.11.9/envs/koku-dev-3.11.9
3.11.9/envs/nise-populator-3.11.9
3.12.0/envs/ansible-core-3.12.0
3.12.0/envs/cla-dev-3.12.0
3.12.0/envs/ipython-3.12.0
3.12.0/envs/jupyter
3.12.0/envs/koku-metrics-operator-3.12.0
3.12.0/envs/pytorch
3.12.0/envs/requests-3.12.0
3.12.0/envs/rh-pre-commit-3.12.0
3.12.1/envs/ipython-3.12.1
3.12.2/envs/1brc
3.12.2/envs/ansible-core-3.12.2
3.12.2/envs/black
3.12.2/envs/cffi-dev-3.12.2
3.12.2/envs/codecov-cli
3.12.2/envs/dr-content-3.12.2
3.12.2/envs/harlequin
3.12.2/envs/ipython-3.12.2
3.12.2/envs/pip-audit
3.12.2/envs/pygments-3.12.2
3.12.2/envs/pyright
3.12.2/envs/qrcode
3.12.2/envs/ruff
3.12.3/envs/ipython-3.12.3
3.12.5/envs/ipython-3.12.5
3.12.5/envs/shellai
3.12.6/envs/build-definitions
3.12.6/envs/ipython-3.12.6
3.12.6/envs/kflux-rel-data
3.12.7/envs/aoc-2024
3.12.7/envs/ipython-3.12.7
3.12.8/envs/dr-backend-3.12.8
3.12.8/envs/ipython-3.12.8
3.12.8/envs/lifecycle-defs
3.12.8/envs/rlsapi-dev
3.12.8/envs/rlsrag
3.13.0/envs/cffi-3.13.0
3.13.0/envs/cicd-tools
3.13.0/envs/ipython-3.13.0
3.13.0/envs/thefuck
3.13.1/envs/docling
3.13.1/envs/httpie
3.13.1/envs/ipython-3.13.1
3.13.1/envs/llm
3.13.1/envs/locust
3.13.1/envs/poetry
3.13.1/envs/prefix
3.13.1/envs/tox
3.6.15/envs/ipython-3.6.15
3.7.15/envs/ipython-3.7.15
3.7.16/envs/ipython-3.7.16
3.8.14/envs/ansible-dev-3.8.14
3.8.15/envs/ipython-3.8.15
3.8.15/envs/nise-populator-3.8.15
3.8.16/envs/koku-nise-dev-3.8.16
3.8.17/envs/ipython-3.8.17
3.8.18/envs/parquet-compactor-3.8.18
3.8.20/envs/ipython-3.8.20
3.9.14/envs/iqe-3.9.14
3.9.15/envs/ipython-3.9.15
3.9.16/envs/koku-dev-logging-3.9.16
3.9.17/envs/event-sign-3.9.17
3.9.17/envs/ipython-3.9.17
3.9.18/envs/koku-dev-3.9.18
3.9.18/envs/koku-rediscache-debug
3.9.18/envs/parquet-compactor-3.9.18
3.9.19/envs/ipython-3.9.19
3.9.19/envs/koku-dev-3.9.19
1brc
ansible-core-3.11.6
ansible-core-3.12.0
ansible-core-3.12.2
ansible-dev-3.11.3
ansible-dev-3.8.14
aoc-2024
argcomplete-3.11.0
asciinema-3.11.0
black
bonfire-3.11.2
borg-3.11.0
borg-3.11.1
build-definitions
cffi-3.13.0
cffi-dev-3.11.8
cffi-dev-3.12.2
cherry_picker-3.10.7
cherry_picker-3.11.0
cicd-tools
cla-dev-3.12.0
codecov-cli
cookiecutter-3.11.0
diceware-3.11.0
docling
dr-backend-3.12.8
dr-content-3.12.2
event-sign-3.9.17
flake8-3.11.0
flash
harlequin
httpie
instruct-lab
ipdb-3.11.0
ipython-3.10.14
ipython-3.10.15
ipython-3.11.10
ipython-3.11.9
ipython-3.12.0
ipython-3.12.1
ipython-3.12.2
ipython-3.12.3
ipython-3.12.5
ipython-3.12.6
ipython-3.12.7
ipython-3.12.8
ipython-3.13.0
ipython-3.13.1
ipython-3.6.15
ipython-3.7.15
ipython-3.7.16
ipython-3.8.15
ipython-3.8.17
ipython-3.8.20
ipython-3.9.15
ipython-3.9.17
ipython-3.9.19
iqe-3.9.14
isort-3.11.0
jello-3.11.0
json-linter-3.11.4
jupyter
kflux-rel-data
koku-dev-3.11.10
koku-dev-3.11.8
koku-dev-3.11.9
koku-dev-3.9.18
koku-dev-3.9.19
koku-dev-dj4-3.11.8
koku-dev-logging-3.9.16
koku-metrics-operator-3.12.0
koku-nise-dev-3.8.16
koku-rediscache-debug
koku-test
lifecycle-defs
llm
locust
mackup-3.11.0
maestral-3.11.0
mitmproxy-3.11.0
mypy-3.11.0
nise-populator-3.11.9
nise-populator-3.8.15
parquet-compactor-3.10.13
parquet-compactor-3.11.5
parquet-compactor-3.8.18
parquet-compactor-3.9.18
pip-audit
pipenv-3.11.7
pipfile2req-3.11.0
poetry
pre-commit
prefix
pur-3.11.0
pygments-3.12.2
pylibssh-dev-3.11.6
pylint-3.11.0
pyls-3.10.9
pyright
pyspy-3.11.0
pytest-3.11.0
pytorch
qontract-3.11.1
qr-coding-3.11.3
qrcode
qrimg-3.11.0
requests-3.11.1
requests-3.12.0
rh-pre-commit-3.12.0
rlsapi-dev
rlsrag
ruff
shellai
shiv-3.11.0
speedtest-cli-3.11.0
sphinx-3.11.0
tasmota-3.11.6
tasmota-3.11.7
testing
thefuck
titlecase-3.11.1
tox
tox-3.11.0
tweet-delete-3.11.0
uptime-kuma-3.11.1
yamllint-3.11.0
youtube-dl-3.11.0
youtube-dl-dev

```
</details>

### Output formatting changes

This PR also includes changes to the output format to make it more consistent with `pyenv versions`.

The completions for `pyenv virtualenv` now omit envs.

samdoran committed:
Improve performance of pyenv-virtualenvs

Use the code from pyenv-versions for efficiency and consistent output.

The main performance problem was in the call to pyenv-virtualenv-prefix,
which called pyenv-prefix, which then enumerated every virtual environment.
This was done inside a loop, compounding the problem.

Simply the virtual environment listing so that it does not have to call
pyenv-virtualenv-prefix anymore.

samdoran committed:
Shellcheck fixes, mostly quoting to avoid word splitting

samdoran committed:
Add --only-aliases argument

This is useful for listing the “frinedly” virtualenv names instead of
the short path which includes the Python version.

samdoran committed:
Only show Python versions for completion when creating new virtual environments

Having the current virtual environments listed as options in the competion
is noisy since only bare Python versions, such as 3.11.1, make sense as
suggested completions for `pyenv virtualenv [version]`.

samdoran committed:
Only show aliases as completions for `pyenv activate`

This makes the suggested completetions cleaner.

samdoran committed:
Update tests

samdoran commented:
I could use some help getting the tests correct. I made a few changes but I don't understand why the `unstub` commands were failing or how to get a virtual env to show up as "activated" in the test environment.

native-api commented:
> With 128 virtual environments, it currently takes about 13 seconds to run

One of the outputs seems wrong :slightly_smiling_face:

samdoran commented:
I pasted the wrong one at first. Sorry about that. 😊

ChristianFredrikJohnsen commented:
Seems like only two tests are failing 😄

I just randomly checked into virtualenv. I recently did some perf improvements on the main `pyenv` repo, and I found myself spending more work fixing the tests (and understanding bats, etc.) than actually improving the source code.
In the `pyenv` repo, I found that some parts of the code were just written to enable tests to pass, and had no actual value beyond that.

I am not sure if this applies to your situation as well, haven't looked much into it. The main idea I have is:
> The code may be perfectly fine, but the tests can be a mess!

samdoran commented:
The tests do need some adjusting since I've changed the output a bit. I'm just having a hard time figuring out what exactly `stub` and `unstub` do as well as how to "activate" a venv in the test environment. Testing is always the hard part.

ChristianFredrikJohnsen commented:
Hehe, and the test file hasn't been touched in 10 years 😆
Same thing I ran into in the `pyenv` project. Probably only the guy who initially created this stuff knows.

I did not encounter `stub` and `unstub` when I did testing in `pyenv`, and there was no `activate` there as well (since no virtualenvs of course). Yeah, I don't have any quick answers 💀 It's a pain 🥲

samdoran commented:
The code itself that was the source of the performance issue is 10-12 years old. Not that old code is bad, but there were some inefficiencies in there.

ChristianFredrikJohnsen commented:
No it's not that the old code is bad. It was probably good at the time it was written. It's more of a problem that the code has design principles which made sense back then, but fast forward 10 years and it's not as intuitive anymore.

And if you're unlucky, you're facing tests which were *quick-fix* set up 10 years ago, and suddenly you need to fix the problem of ...
Not necessarily the problem of any one person either, it's just that the entire structure doesn't feel intuitive.

It's a challenge for sure 😄

native-api commented on test/virtualenvs.bats:
```diff
@@ -21,40 +26,40 @@ setup() {
   stub pyenv-virtualenv-prefix "3.3.3 : false"
-  stub pyenv-virtualenv-prefix "venv27 : echo \"/usr\""
-  stub pyenv-virtualenv-prefix "venv33 : echo \"/usr\""
+  stub pyenv-virtualenv-prefix "venv27 : echo \"${PYENV_ROOT}/versions/2.7.6\""
+  stub pyenv-virtualenv-prefix "venv33 : echo \"${PYENV_ROOT}/versions/3.3.3\""
```
Not sure if this change improves anything. As a rule of thumb, one should not change test cases unless you've changed the behavior in them -- to avoid accidentally losing test cover.

native-api commented on test/virtualenvs.bats:
```diff
@@ -21,40 +26,40 @@ setup() {
 
-  unstub pyenv-version-name
-  unstub pyenv-virtualenv-prefix
+  # unstub pyenv-version-name
```
`unstub` checcks if the specific stubbed executable was called exactly according to the sequence set up with `stub` calls. If not, it fails and prints the plan and the fact.
The implementation is in `test_helper.bash` and the `stubs/stub` script which the former calls.

samdoran replied:
I read through this some more and I have a better understanding of what those helpers are doing. They should be correct in the tests now.

native-api commented on bin/pyenv-activate:
```diff
@@ -17,7 +17,7 @@ set -e
 if [ "$1" = "--complete" ]; then
   echo --unset
-  exec pyenv-virtualenvs --bare
+  exec pyenv-virtualenvs --bare --only-aliases
```
This is a design change -- so this should be discussed separately. Envs are supposed to be accessible under both full name and alias. I'm not sure why that is; aliases are internally called "compat", indicating that they were added for compatibility with... something.

samdoran replied:
I'll remove that change for now and we can discuss elsewhere. What's the best place to have a design discussion?

native-api replied:
> What's the best place to have a design discussion?

Discussions.

native-api commented on bin/pyenv-sh-activate:
```diff
@@ -51,7 +51,7 @@ while [ $# -gt 0 ]; do
     # Provide pyenv completions
     echo --unset
-    exec pyenv-virtualenvs --bare
+    exec pyenv-virtualenvs --bare --only-aliases
```
same as above

samdoran commented:
Thank you for the feedback @native-api. I'll address this later in the week.

samdoran commented:
I had a crazy week and a full weekend away from home. I will work on this during the week.

samdoran commented on bin/pyenv-virtualenvs:
```diff
@@ -34,10 +34,11 @@ for arg; do
   case "$arg" in
   --complete )
     echo --bare
-    echo --skip-aliases
```
Oops. I meant to add a new option not remove that one.

samdoran committed:
Add skip-aliases to complete output

This was removed by accident.

samdoran commented on bin/pyenv-virtualenvs:
```diff
@@ -82,39 +119,73 @@ exists() {
-    echo "${miss_prefix}${1}${print_origin+$2}"
+    version_repr="$version"
+  fi
+  if [[ ${BASH_VERSINFO[0]} -ge 4 && ${current_versions["$1"]} ]]; then
```
I had to look this up. `((expression))` evaluates according to the Arithmetic Evaluation rules, meaning comparison operators such as `<` and `>` are allowed.

`[[ expression ]]` evaluates according to the Conditional Expressions rules, meaning path expansion and word splitting happen and `-lt` and `-gt` operators are required when comparing integers.

It seems `[[ expression ]]` is the preferred approach these days so I'll use that for consistency.

samdoran committed:
Use conditional expressions

For consistency, use conditional expressions instead of
arithmetic evaluation when comparing bash versions.

samdoran commented on bin/pyenv-virtualenvs:
```diff
@@ -47,24 +48,60 @@ done
 
-if [ -d "$versions_dir" ]; then
-  versions_dir="$(realpath "$versions_dir")"
+if ! enable -f "${BASH_SOURCE%/*}"/pyenv-realpath.dylib realpath 2>/dev/null; then
+  if [ -n "$PYENV_NATIVE_EXT" ]; then
+    echo "pyenv: failed to load \`realpath' builtin" >&2
+    exit 1
+  fi
+
+  READLINK=$(type -P readlink)
+  if [ -z "$READLINK" ]; then
+    echo "pyenv: cannot find readlink - are you missing GNU coreutils?" >&2
+    exit 1
+  fi
+
+  resolve_link() {
+    $READLINK "$1"
+  }
+
+  realpath() {
+    local path="$1"
+    local name
+
+    # Use a subshell to avoid changing the current path
+    (
+    while [ -n "$path" ]; do
+      name="${path##*/}"
+      [ "$name" = "$path" ] || cd "${path%/*}"
+      path="$(resolve_link "$name" || true)"
```
Good catch. This is code that I moved over from `pyenv-versions`. The changes in this PR avoid the use of `realpath` altogether, so this can all be removed.

samdoran committed:
Remove unused code

The new implementation is not using realpath.

samdoran committed:
Use the newer [[ command

https://mywiki.wooledge.org/BashFAQ/031

samdoran commented on bin/pyenv-virtualenvs:
```diff
@@ -82,39 +119,73 @@ exists() {
+fi
+
+if [ -z "$skip_aliases" ]; then
+  for env_path in "${version_dir_entries[@]}"; do
+    if [ -d "${env_path}" ] && [ -L "${env_path}" ]; then
+      print_version "${env_path#"${PYENV_ROOT}"/versions/}" "${env_path}"
     fi
-    for venv_path in "${path}/envs/"*; do
-      venv="${path##*/}/envs/${venv_path##*/}"
-      virtualenv_prefix="$(pyenv-virtualenv-prefix "${venv}" 2>/dev/null || true)"
-      if [ -d "${virtualenv_prefix}" ]; then
-        print_version "${venv}" " (created from ${virtualenv_prefix})"
-      fi
-    done
-  fi
-done
+  done
+fi
```
I don't know anything about or use conda, so I omitted those. 😊
I will work on adding that back in as they should be listed in the output.

samdoran replied:
Reading through the previous code, it seems like the intent was to look for certain indicator files and if they are present, list the directory as a virtual environment. I don't really know if that's a good heuristic or not.

Calling `pyenv-virtualenv-prefix` was what caused the program to take so long to run. If we think this behavior should remain, I'll have to reimplement this functionality using a more efficient approach.

samdoran replied:
Hmm, the changes in this PR may actually be doing a better job with this than the current code.

I installed `miniconda3` and created a new env.

```
pyenv install miniconda3-latest
pyenv shell miniconda3-latest
conda create -n my_conda_env
```

I only see `my_conda_env` in the second output.

<details>
  <summary>Current output running <code>master</code></summary>
  
```
> pyenv virtualenvs
  3.10.2/envs/pyls-3.10.2 (created from /Users/sdoran/.pyenv/versions/3.10.2)
  3.10.3/envs/ansible-2.12-3.10.3 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  3.10.3/envs/boto-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  3.10.3/envs/lexicon-3.10.3 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  3.10.4/envs/ansible-2.12-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.4/envs/ansible-dev-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.4/envs/flake8-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.4/envs/jello-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.4/envs/jsbeautifier-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.4/envs/youtube-dl-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  3.10.6/envs/ipython-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  3.10.6/envs/shiv-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  3.10.6/envs/speedtest-cli-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  3.11.1/envs/ansible-2.14-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  3.11.1/envs/ipython-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  3.11.1/envs/titlecase-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  3.11.1/envs/uptime-kuma-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  3.11.3/envs/ipython-3.11.3 (created from /Users/sdoran/.pyenv/versions/3.11.3)
  3.11.3/envs/tasmota (created from /Users/sdoran/.pyenv/versions/3.11.3)
  3.11.7/envs/tasmota-3.11.7 (created from /Users/sdoran/.pyenv/versions/3.11.7)
  3.12.0/envs/ansible-2.16-3.12.0 (created from /Users/sdoran/.pyenv/versions/3.12.0)
  3.12.0/envs/ipython-3.12.0 (created from /Users/sdoran/.pyenv/versions/3.12.0)
  3.12.2/envs/borg (created from /Users/sdoran/.pyenv/versions/3.12.2)
  3.12.2/envs/yt-dlp (created from /Users/sdoran/.pyenv/versions/3.12.2)
  3.12.4/envs/qrcode (created from /Users/sdoran/.pyenv/versions/3.12.4)
  3.12.8/envs/ansible-lint (created from /Users/sdoran/.pyenv/versions/3.12.8)
  3.12.8/envs/cowsay (created from /Users/sdoran/.pyenv/versions/3.12.8)
  3.12.8/envs/pyright-3.12.8 (created from /Users/sdoran/.pyenv/versions/3.12.8)
  3.12.8/envs/ruff-3.12.8 (created from /Users/sdoran/.pyenv/versions/3.12.8)
  3.13.1/envs/asciinema (created from /Users/sdoran/.pyenv/versions/3.13.1)
  3.13.1/envs/diceware (created from /Users/sdoran/.pyenv/versions/3.13.1)
  3.13.1/envs/ipython-3.13.1 (created from /Users/sdoran/.pyenv/versions/3.13.1)
  3.6.15/envs/ipython-3.6.15 (created from /Users/sdoran/.pyenv/versions/3.6.15)
  3.7.12/envs/appomni-dev-3.7.12 (created from /Users/sdoran/.pyenv/versions/3.7.12)
  3.7.12/envs/ipython-3.7.12 (created from /Users/sdoran/.pyenv/versions/3.7.12)
  3.8.12/envs/ipython-3.8.12 (created from /Users/sdoran/.pyenv/versions/3.8.12)
  3.9.9/envs/ipython-3.9.9 (created from /Users/sdoran/.pyenv/versions/3.9.9)
  ansible-2.12-3.10.3 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  ansible-2.12-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  ansible-2.14-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  ansible-2.16-3.12.0 (created from /Users/sdoran/.pyenv/versions/3.12.0)
  ansible-dev-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  ansible-lint (created from /Users/sdoran/.pyenv/versions/3.12.8)
  appomni-dev-3.7.12 (created from /Users/sdoran/.pyenv/versions/3.7.12)
  asciinema (created from /Users/sdoran/.pyenv/versions/3.13.1)
  borg (created from /Users/sdoran/.pyenv/versions/3.12.2)
  boto-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  cowsay (created from /Users/sdoran/.pyenv/versions/3.12.8)
  diceware (created from /Users/sdoran/.pyenv/versions/3.13.1)
  flake8-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  ipython-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  ipython-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  ipython-3.11.3 (created from /Users/sdoran/.pyenv/versions/3.11.3)
  ipython-3.12.0 (created from /Users/sdoran/.pyenv/versions/3.12.0)
  ipython-3.13.1 (created from /Users/sdoran/.pyenv/versions/3.13.1)
  ipython-3.6.15 (created from /Users/sdoran/.pyenv/versions/3.6.15)
  ipython-3.7.12 (created from /Users/sdoran/.pyenv/versions/3.7.12)
  ipython-3.8.12 (created from /Users/sdoran/.pyenv/versions/3.8.12)
  ipython-3.9.9 (created from /Users/sdoran/.pyenv/versions/3.9.9)
  jello-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  jsbeautifier-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  lexicon-3.10.3 (created from /Users/sdoran/.pyenv/versions/3.10.3)
  miniconda3-latest (created from /Users/sdoran/.pyenv/versions/miniconda3-latest)
  pyls-3.10.2 (created from /Users/sdoran/.pyenv/versions/3.10.2)
  pyright-3.12.8 (created from /Users/sdoran/.pyenv/versions/3.12.8)
  qrcode (created from /Users/sdoran/.pyenv/versions/3.12.4)
  ruff-3.12.8 (created from /Users/sdoran/.pyenv/versions/3.12.8)
  shiv-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  speedtest-cli-3.10.6 (created from /Users/sdoran/.pyenv/versions/3.10.6)
  tasmota (created from /Users/sdoran/.pyenv/versions/3.11.3)
  tasmota-3.11.7 (created from /Users/sdoran/.pyenv/versions/3.11.7)
  titlecase-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  uptime-kuma-3.11.1 (created from /Users/sdoran/.pyenv/versions/3.11.1)
  youtube-dl-3.10.4 (created from /Users/sdoran/.pyenv/versions/3.10.4)
  yt-dlp (created from /Users/sdoran/.pyenv/versions/3.12.2)
```
</details>

<details>
  <summary>New output from this branch</summary>
  
```
  3.10.2/envs/pyls-3.10.2
  3.10.3/envs/ansible-2.12-3.10.3
  3.10.3/envs/boto-3.10.4
  3.10.3/envs/lexicon-3.10.3
  3.10.4/envs/ansible-2.12-3.10.4
  3.10.4/envs/ansible-dev-3.10.4
  3.10.4/envs/flake8-3.10.4
  3.10.4/envs/jello-3.10.4
  3.10.4/envs/jsbeautifier-3.10.4
  3.10.4/envs/youtube-dl-3.10.4
  3.10.6/envs/ipython-3.10.6
  3.10.6/envs/shiv-3.10.6
  3.10.6/envs/speedtest-cli-3.10.6
  3.11.1/envs/ansible-2.14-3.11.1
  3.11.1/envs/ipython-3.11.1
  3.11.1/envs/titlecase-3.11.1
  3.11.1/envs/uptime-kuma-3.11.1
  3.11.3/envs/ipython-3.11.3
  3.11.3/envs/tasmota
  3.11.7/envs/tasmota-3.11.7
  3.12.0/envs/ansible-2.16-3.12.0
  3.12.0/envs/ipython-3.12.0
  3.12.2/envs/borg
  3.12.2/envs/yt-dlp
  3.12.4/envs/qrcode
  3.12.8/envs/ansible-lint
  3.12.8/envs/cowsay
  3.12.8/envs/pyright-3.12.8
  3.12.8/envs/ruff-3.12.8
  3.13.1/envs/asciinema
  3.13.1/envs/diceware
  3.13.1/envs/ipython-3.13.1
  3.6.15/envs/ipython-3.6.15
  3.7.12/envs/appomni-dev-3.7.12
  3.7.12/envs/ipython-3.7.12
  3.8.12/envs/ipython-3.8.12
  3.9.9/envs/ipython-3.9.9
  miniconda3-latest/envs/my_conda_env
  ansible-2.12-3.10.3 --> /Users/sdoran/.pyenv/versions/3.10.3/envs/ansible-2.12-3.10.3
  ansible-2.12-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/ansible-2.12-3.10.4
  ansible-2.14-3.11.1 --> /Users/sdoran/.pyenv/versions/3.11.1/envs/ansible-2.14-3.11.1
  ansible-2.16-3.12.0 --> /Users/sdoran/.pyenv/versions/3.12.0/envs/ansible-2.16-3.12.0
  ansible-dev-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/ansible-dev-3.10.4
  ansible-lint --> /Users/sdoran/.pyenv/versions/3.12.8/envs/ansible-lint
  appomni-dev-3.7.12 --> /Users/sdoran/.pyenv/versions/3.7.12/envs/appomni-dev-3.7.12
  asciinema --> /Users/sdoran/.pyenv/versions/3.13.1/envs/asciinema
  borg --> /Users/sdoran/.pyenv/versions/3.12.2/envs/borg
  boto-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.3/envs/boto-3.10.4
  cowsay --> /Users/sdoran/.pyenv/versions/3.12.8/envs/cowsay
  diceware --> /Users/sdoran/.pyenv/versions/3.13.1/envs/diceware
  flake8-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/flake8-3.10.4
  ipython-3.10.6 --> /Users/sdoran/.pyenv/versions/3.10.6/envs/ipython-3.10.6
  ipython-3.11.1 --> /Users/sdoran/.pyenv/versions/3.11.1/envs/ipython-3.11.1
  ipython-3.11.3 --> /Users/sdoran/.pyenv/versions/3.11.3/envs/ipython-3.11.3
  ipython-3.12.0 --> /Users/sdoran/.pyenv/versions/3.12.0/envs/ipython-3.12.0
  ipython-3.13.1 --> /Users/sdoran/.pyenv/versions/3.13.1/envs/ipython-3.13.1
  ipython-3.6.15 --> /Users/sdoran/.pyenv/versions/3.6.15/envs/ipython-3.6.15
  ipython-3.7.12 --> /Users/sdoran/.pyenv/versions/3.7.12/envs/ipython-3.7.12
  ipython-3.8.12 --> /Users/sdoran/.pyenv/versions/3.8.12/envs/ipython-3.8.12
  ipython-3.9.9 --> /Users/sdoran/.pyenv/versions/3.9.9/envs/ipython-3.9.9
  jello-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/jello-3.10.4
  jsbeautifier-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/jsbeautifier-3.10.4
  lexicon-3.10.3 --> /Users/sdoran/.pyenv/versions/3.10.3/envs/lexicon-3.10.3
  pyls-3.10.2 --> /Users/sdoran/.pyenv/versions/3.10.2/envs/pyls-3.10.2
  pyright-3.12.8 --> /Users/sdoran/.pyenv/versions/3.12.8/envs/pyright-3.12.8
  qrcode --> /Users/sdoran/.pyenv/versions/3.12.4/envs/qrcode
  ruff-3.12.8 --> /Users/sdoran/.pyenv/versions/3.12.8/envs/ruff-3.12.8
  shiv-3.10.6 --> /Users/sdoran/.pyenv/versions/3.10.6/envs/shiv-3.10.6
  speedtest-cli-3.10.6 --> /Users/sdoran/.pyenv/versions/3.10.6/envs/speedtest-cli-3.10.6
  tasmota --> /Users/sdoran/.pyenv/versions/3.11.3/envs/tasmota
  tasmota-3.11.7 --> /Users/sdoran/.pyenv/versions/3.11.7/envs/tasmota-3.11.7
  titlecase-3.11.1 --> /Users/sdoran/.pyenv/versions/3.11.1/envs/titlecase-3.11.1
  uptime-kuma-3.11.1 --> /Users/sdoran/.pyenv/versions/3.11.1/envs/uptime-kuma-3.11.1
  youtube-dl-3.10.4 --> /Users/sdoran/.pyenv/versions/3.10.4/envs/youtube-dl-3.10.4
  yt-dlp --> /Users/sdoran/.pyenv/versions/3.12.2/envs/yt-dlp

```
</details>

samdoran replied:
I just double checked this using a simpler `PYENV_ROOT` and `conda` environments are listed as expected.

```
> set -x PYENV_ROOT /tmp/pyenv
> pyenv install miniconda3-3.12-24.1.2-0
> pyenv install 3.14.4
> pyenv shell miniconda3-3.12-24.1.2-0
> pyenv versions
  system
  3.14.4
* miniconda3-3.12-24.1.2-0 (set by PYENV_VERSION environment variable)

> conda create -n my_conda_env
> pyenv virtualenvs
  miniconda3-3.12-24.1.2-0/envs/my_conda_env

> pyenv virtualenvs --skip-aliases
  miniconda3-3.12-24.1.2-0/envs/my_conda_env

> pyenv virtualenvs --bare
miniconda3-3.12-24.1.2-0/envs/my_conda_env
```

samdoran committed:
Remove --only-aliases option for later discussion

samdoran commented:
Thanks for all the feedback. I didn't get to fixing up the tests. I need to get some sleep for now.

samdoran commented:
I have not forgotten about this. My day job has consumed all my capacity lately so I haven't circled back to address the test failures.

I'm kinda cheating because I started using the changes in this branch locally so haven't felt the pain in a while.

samdoran committed:
Correct test setup so aliases are properly symlinked

samdoran committed:
Update tests and add more test cases

Update the existing tests based on new behavior/output format.
Add additional tests to cover scenarios that were not previously tested.

samdoran commented on test/virtualenvs.bats:
```diff
@@ -4,10 +4,15 @@ load test_helper
+  mkdir -p "${PYENV_ROOT}/versions/2.7.6/envs/venv27"
+  mkdir -p "${PYENV_ROOT}/versions/3.3.3/envs/venv33"
+  ln -s "venv27" "${PYENV_ROOT}/versions/venv27"
+  ln -s "venv33" "${PYENV_ROOT}/versions/venv33"
```
Yup, that was incorrect and caused me a lot of confusion. I _finally_ fixed that and the tests are better behaved now.

native-api commented on bin/pyenv-virtualenv:
```diff
@@ -24,7 +24,7 @@ fi
 # Provide pyenv completions
 if [ "$1" = "--complete" ]; then
-  exec pyenv-versions --bare
+  exec pyenv-versions --bare --skip-envs --skip-aliases
```
`--skip-envs` seems logical -- neither `venv` not `virtualenv` allows to chain virtualenvs IIRC.

What's the rationale for `--skip-aliases` though?

samdoran replied:
Without `--skip-aliases`, current virtual environments show up in completions. I didn't know you could create a new venv from an existing one, so it always looked wrong including anything other than the installed Python versions.

It seems it is actually possible to create a new venv using a current one as the `[version]`. While I prefer the cleaner completions without envs or aliases, it would probably be a a regression.

native-api replied:
> Without --skip-aliases, current virtual environments show up in completions.

I don't think they should... need to check the documentation.
This is something to fix in `pyenv-versions`.

native-api replied:
> > Without --skip-aliases, current virtual environments show up in completions.
> 
> I don't think they should... need to check the documentation. This is something to fix in `pyenv-versions`.

Darn. The documentation does say exactly how it is now. So that'll be an interface change. Probably worth having a discussion to get user feedback.

native-api committed:
Revert `activate` completion hint change

native-api committed:
Refactor

native-api commented on bin/pyenv-virtualenvs:
```diff
@@ -82,39 +70,70 @@ exists() {
 shopt -u nullglob
 
-if [ "$num_versions" -eq 0 ] && [ -n "$include_system" ]; then
+if [[ $num_versions -eq 0 ]] && [[ -n $include_system ]]; then
   echo "Warning: no Python virtualenv detected on the system" >&2
   exit 1
 fi
```
Do we really need a warning if there are no virtual environments? That's a normal situation.

AFAICS, in the original version, `include_system` was never set. So this is dead code, likely an artifact from the previous round of copying from `pyenv-versions` 🙂 .

samdoran replied:
Yup, looks like the warning never showed up in the original code. I'll just remove this because it does seem like noise.

samdoran committed:
Include aliases in completetions

While it looks a bit funny, it is valid to use an existing virtual
environment as the [version] when creating a new one.

samdoran committed:
Remove warning for no virtualenvs

native-api commented on test/virtualenvs.bats:
```diff
@@ -4,63 +4,116 @@ load test_helper
 
   assert_success
   assert_output <<OUT
-  venv27 (created from /usr)
-* venv33 (created from /usr)
```
Base environment is critical information for an env. Unless you have rationale for it, it should stay.

samdoran replied:
Calling `pyenv-virtualenv-prefix` in order to get that information is a huge source of the performance problem. I'll see if I can find a more efficient way of getting that information.

samdoran replied:
Actually, I think that is _the_ source of the performance problems. With 160 virtual environments, calling `pyenv-virtualenv-prefix` to get the base environment increases the total output time from 665 ms to 12 s.

```
> pyenv virtualenvs --skip-aliases | wc -l
     160

> time pyenv virtualenvs > /dev/null 2>&1

________________________________________________________
Executed in   12.73 secs    fish           external
   usr time    5.99 secs    0.20 millis    5.99 secs
   sys time    6.11 secs    1.14 millis    6.11 secs

# omitting the base environment lookup with pyenv-virtualenv-prefix
> time pyenv virtualenvs > /dev/null 2>&1

________________________________________________________
Executed in  664.70 millis    fish           external
   usr time  168.56 millis  189.00 micros  168.37 millis
   sys time  195.23 millis  889.00 micros  194.34 millis
```

native-api replied:
I see that `pyenv-virtualenv-prefix` accepts multiple arguments. Would calling it just once help?

samdoran replied:
It looks like I can eliminate most of the performance problem by omitting the call to `pyenv-prefix "${version}"` in `pyenv-virtualenv-prefix`. `pyenv-prefix` is what ends up iterating over everything in `versions/`, making this O(N^2).

There are a few more improvements that can be made with the spelunking done in `pyenv-virtualenv-prefix`.

native-api replied:
> `pyenv-prefix` is what ends up iterating over everything in `versions/`, making this O(N^2).

I think we can fix that by making a fast path in `pyenv-latest` for when there is an exact match -- which is going to be virtually all cases.

native-api replied:
> > `pyenv-prefix` is what ends up iterating over everything in `versions/`, making this O(N^2).
> 
> I think we can fix that by making a fast path in `pyenv-latest` for when there is an exact match -- which is going to be virtually all cases.

https://github.com/native-api/pyenv/tree/latest_fast_path

samdoran replied:
What I came up with was just setting `PYENV_PREFIX_PATH="${PYENV_ROOT}/versions/${version}"` in `pyenv-virtualenv-prefix`, avoiding the call to `pyenv-prefix`. Would that be ok or should be still check for the directory?

samdoran replied:
Your change does help. The baseline without any changes is 13.98s.

With the changes in your branch:
```
> time pyenv virtualenvs > /dev/null 2>&1

________________________________________________________
Executed in    5.42 secs    fish           external
   usr time    2.48 secs  207.00 micros    2.48 secs
   sys time    2.25 secs  758.00 micros    2.25 secs
```

If I move that check into `pyenv-virtualenv-prefix`, it cuts off about 2.8 more seconds.
```
> time pyenv virtualenvs > /dev/null 2>&1

________________________________________________________
Executed in    2.60 secs    fish           external
   usr time    1.20 secs  191.00 micros    1.20 secs
   sys time    1.27 secs  747.00 micros    1.27 secs
```

I'm going to push that up.

native-api replied:
> What I came up with was just setting `PYENV_PREFIX_PATH="${PYENV_ROOT}/versions/${version}"` in `pyenv-virtualenv-prefix`, avoiding the call to `pyenv-prefix`. Would that be ok or should be still check for the directory?

Currently, `pyenv-virtualenv-prefix` accepts any version -- even unresolved ones that would go through `pyenv-latest`. And without arguments, it takes the currently selected version(s) -- which can very well contain an unresolved one.

So it's only safe to cut the `pyenv-prefix` call if we "know" the version arguments are already resolved.
That is the case for calls from `pyenv-virtualenvs` -- so we can e,g, introduce an internal command line switch to indicate that.

native-api replied:
> Your change does help. The baseline without any changes is 13.98s.
> 
> With the changes in your branch:
> 
> ```
> > time pyenv virtualenvs > /dev/null 2>&1
> 
> ________________________________________________________
> Executed in    5.42 secs    fish           external
>    usr time    2.48 secs  207.00 micros    2.48 secs
>    sys time    2.25 secs  758.00 micros    2.25 secs
> ```

Okay -- [in that change goes](https://github.com/pyenv/pyenv/pull/3437)!

samdoran replied:
Sweet!

samdoran committed:
Check for existence of venv directory

Only call pyenv-prefix if the path doesn’t exist.

samdoran committed:
Fix incorrect IFS handling

The echo "${VIRTUALENV_PREFIX_PATHS[*]}" statement at the end of the file
only worked by accident because IFS=: was set earlier in the file.

Properly set/unset IFS.

samdoran commented:
Doh, let me see what's up with the test failures.

samdoran committed:
Remove pyenv-prefix stubs from tests

This refelects how call changes have changed in the code.

native-api committed:
Add a test for version resolution

native-api committed:
fix test

in case of path with spaces

native-api commented on test/virtualenvs.bats:
```diff
@@ -4,63 +4,119 @@ load test_helper
 
   assert_success
   assert_output <<OUT
-  venv27 (created from /usr)
-* venv33 (created from /usr)
```
Bug: pyenv-virtualenvs now ignores envs that are based on system.
```
$ sudo apt install python3.12-venv
<...>
$ pyenv virtualenv test_system
<...>
$ ls -l ~/.pyenv/versions/
total 8
drwxr-xr-x 7 vmuser vmuser 4096 Apr 21 08:06 3.14.3
lrwxrwxrwx 1 vmuser vmuser   45 Apr 23 23:53 test -> /home/vmuser/.pyenv/versions/3.14.3/envs/test
drwxrwxr-x 5 vmuser vmuser 4096 Apr 24 16:58 test_system
lrwxrwxrwx 1 vmuser vmuser   43 Apr 21 08:06 vt -> /home/vmuser/.pyenv/versions/3.14.3/envs/vt
$ pyenv virtualenvs
  3.14.3/envs/test (created from /home/vmuser/.pyenv/versions/3.14.3)
  3.14.3/envs/vt (created from /home/vmuser/.pyenv/versions/3.14.3)
  test --> /home/vmuser/.pyenv/versions/3.14.3/envs/test
  vt --> /home/vmuser/.pyenv/versions/3.14.3/envs/vt
```
Looks like messing with existing test data without knowing what we're doing wasn't a good idea.

native-api replied:
I guess the easiest way would be to revert the stock test data to what it was and work from there.
E.g. we probably want to add one alias entry to the stock data -- so that we have envs both with and without an alias.
It also makes sense to e.g. refactor stock data out of `setup()` if it's not needed in all cases, or add a function to remove it if it's only not needed in very few cases.

samdoran replied:
Good catch. Glad you caught you that.

samdoran replied:
There is definitely a bug that's preventing system venvs from being shown. I'll figure that out.

I did some testing with only creating a venv for the tests that need it instead of having it there for all tests and that seems the work. I'll work on tailoring the setup of each test.

samdoran replied:
I'm still hacking on this but making good progress. I fixed the bug and am just working on tests.

I have a much better understanding of the test machinery now. I need to head out for some family stuff but I'll pick it up later.

samdoran replied:
I made a good bit of changes to the tests and created separate setup functions for system and non-system venvs.  I explained the changes in the commit history to make it easier to follow the thinking behind the changes. There is some redundancy for the sake of clarity. I'm open to feedback since I've been staring at it for a while.

native-api commented on test/virtualenvs.bats:
```diff
@@ -4,63 +4,119 @@ load test_helper
-  mkdir -p "${PYENV_ROOT}/versions/venv33"
+  setup_m_venv "2.7.6/envs/venv27"
+  echo "home = ${PYENV_ROOT}/versions/2.7.6/bin" > "${PYENV_ROOT}/versions/2.7.6/envs/venv27/pyvenv.cfg"
+  ln -s "${PYENV_ROOT}/versions/2.7.6/envs/venv27" "${PYENV_ROOT}/versions/venv27"
```
Probably better refactor this into `create_m_venv()` located in this file.

samdoran replied:
Yup, I agree. I'm starting to get the hang of these tests.

native-api committed:
- dead code

native-api commented on test/virtualenvs.bats:
```diff
@@ -4,63 +4,119 @@ load test_helper
+
+  assert_success
+  assert_output ""
+}
```
Without the code for the error message, this is probably redundant

samdoran committed:
Remove unnecessary test

The warning was removed, so testing no empty output doesn’t make much sense.

samdoran committed:
Add a helper for creating virtual environments

samdoran committed:
Change the interface of create_m_venv to match setup_m_venv

Also have create_m_venv call setup_m_venv.

samdoran committed:
Show venvs created from system Python

Venvs created from system Python are directories, not sym links. The previous
code only displayed venvs that were a directory and a symlink. Now check for
directory existence and symlink status, or directory existence and the presence
of bin/activate within the directory.

samdoran committed:
Remove counter that is no longer used

This was only used by the warning code, which was removed.

samdoran committed:
Add setup function for creating a mock system venv

Change the main setup to only set PYENV_ROOT. Each test now sets up the
required virtual environments. This results in some duplication but makes
it clearer what each test is doing.

An alternative approach would be to modify the venv created by create_m_venv to
make it look like a system venv, but two distinct functions make it much clearer
what they are doing even though there is some overlap between the functions.

samdoran committed:
Force creation of symlink if it exists

This makes test setup more resilient.

samdoran committed:
Add a performance regression test

samdoran committed:
Add a few missed unstub calls

Also stub out calls to pyenv-virtualenv-prefix in the test that’s listing venvs.

samdoran committed:
Remove performance test

Making this an integration test ended up giving mixed results.
This test when run without the performance changes in this branch did
not reflect the performance gains seen when running pyenv-virtualenvs
outside of the test environment. The test was not accomplishing its goal.

native-api committed:
test_prefix.bats: - remove_*

no longer needed now that PYENV_ROOT is under BATS_TEST_DIRNAME

native-api committed:
+ refactor notes
